_posts/2018-01-26-grpc-and-deadlines.md

```diff
@@ -0,0 +1,139 @@
+---
+layout: post
+title: gRPC and Deadlines
+published: 
+permalink: blog/deadlines
+author: Gráinne Sheerin, Google SRE
+company: Google
+company-link: https://www.google.com
+---
+
+**TL;DR Always set a deadline**. This post explains why we recommend being deliberate about setting deadlines, with useful code snippets to show you how.
+
+<!--more-->
+
+When you use gRPC, the gRPC library takes care of communication, marshalling, unmarshalling, and deadline enforcement. Deadlines allow gRPC clients to specify how long they are willing to wait for an RPC to complete before the RPC is terminated with the error `DEADLINE_EXCEEDED`. By default this deadline is a very large number, dependent on the language implementation. How deadlines are specified is also language-dependent. Some language APIs work in terms of a **deadline**, a fixed point in time by which the RPC should complete. Others use a **timeout**, a duration of time after which the RPC times out.
+
+In general, when you don't set a deadline resources will be held for all in-flight requests, and all requests can potentially reach the maximum timeout. This puts the service at risk of running out of resources like memory and crashing, as well as maximizing the latency of the service.
+
+To avoid this, services should specify the longest default deadline they technically support, and clients should wait until the response is no longer useful to them. For the service this can be as simple as providing a comment in the .proto file. For the client this involves setting useful deadlines.
+
+There is no single answer to "What is a good deadline/timeout value?". Your service might be as simple as the Greeter in our quick start guides, in which case 100 ms would be fine. Your service might be as complex as a globally-distributed and strongly consistent database. The deadline for a client query will be different from how long they should wait for you to drop their table.
```

Comment:
Could we please link them to the greeter service in our quickstart? Any one will do.